README
NAME

    Data::QuickMemoPlus::Reader - Extract text from QuickMemo+ LQM export
    files.

SYNOPSIS

        use Data::QuickMemoPlus::Reader qw(lqm_to_str);
        my $memo_text = lqm_to_str('QuickMemo+_191208_220400.lqm');
    
        use Data::QuickMemoPlus::Reader qw(lqm_to_txt);
        my $files_converted1 = lqm_to_txt('QuickMemo+_191208_220400.lqm');
        my $files_converted2 = lqm_to_txt('path/to/lqm_files');
        
        ## Omit the header text by setting setting this package variable to false:
        local $Data::QuickMemoPlus::Reader::IncludeHeader;

DESCRIPTION

    Data::QuickMemoPlus::Reader is a module that will extract the text
    contents from archived QuickMemo+ memos.

    QuickMemo+ lqm files are in Zip format. This module unzips them, parses
    the json file inside, then extracts the category and memo text from the
    Json file.

    If the filename of the lqm file contains the original timestamp then
    that is placed in a text header in the text along with the category
    name. The header can be disabled by setting the package variable
    $IncludeHeader to false.

    The following functions are available:

 lqm_to_txt('directory or filename')

    Creates a text file with the same name as each original lqm file but
    with a txt extension. Return value is the number of files successfully
    converted.

 lqm_to_str('filename')

    Returns the text extracted from the lqm file.

LICENSE

    Copyright (C) Brent Shields.

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

AUTHOR

    Brent Shields <bshields@cpan.org>



Changes
Revision history for Perl extension Data-QuickMemoPlus-Reader

{{$NEXT}}

0.11 2021-02-11T23:49:05Z
         - attempt to fix path in meta files.
0.10 2021-02-11T19:23:50Z
	- added META.yml with fixed path.
0.09 2021-02-11T19:00:43Z
	- removed META.yml, it contained bad path.
0.08 2021-02-10T01:48:47Z
	- second attempt at stable release.
0.07 2021-02-10T01:42:34Z
	- attempt to release as stable.
0.06 2021-02-09T18:44:44Z
	- set status to stable in META.json
0.05 2021-02-09T18:16:04Z
	- set rev in META.yml to 0.05
0.04 2021-02-07T18:57:32Z
	- added to pod
0.03 2021-02-07T17:53:30Z
	- fixed meta.yml.
0.02 2021-02-06T23:02:44Z
	- added lqm_to_text.
0.01 2021-01-31T03:18:27Z
	- updated minimum Perl to 5.010
0.01 2021-01-31T03:03:55Z
	- original version

t/lib/ExampleMemo.pm
package ExampleMemo;

sub memo_with_header { 
    header() . memo_no_header();
}

sub memo_with_header_no_timestamp { 
    header_without_timestamp() . memo_no_header();
}

sub memo_no_header {
<<'END_MEMO';
Group mayhem
Interesting docudrama
Until the lights take us

Beatles across the universe

END_MEMO
}

sub header {
<<'END_HEADER';
Created date: 191208_220400
Category:   My_memos
-------------------------------------------------------------------------------
END_HEADER
}

sub header_without_timestamp {

<<'END_HEADER';
Created date: 
Category:   My_memos
-------------------------------------------------------------------------------
END_HEADER
}

sub jlqm {
    chomp(my $jlqm_contents = <<'END_JLQM');
{
  "Category": {
    "AccountName": "phone",
    "CategoryName": "My memos",
    "OriginalCategoryName": "My memos",
    "Icon": 0,
    "Id": 1,
    "IsDefault": 0,
    "IsSynced": 0,
    "MemoCount": -1,
    "Order": 0
  },
  "Memo": {
    "CheckboxDesc": "",
    "Desc": "\u003cdiv align\u003d\"left\" \u003e\u003cp dir\u003d\"ltr\"\u003e\u003cfont color \u003d\"-16777216\"\u003eGroup mayhem\u003c/font\u003e\u003c/p\u003e\u003c/div\u003e\u003cp dir\u003d\"ltr\"\u003e\u003cbr\u003e\u003c/p\u003e\u003cdiv align\u003d\"left\" \u003e\u003cp dir\u003d\"ltr\"\u003e\u003cfont color \u003d\"-16777216\"\u003eInteresting docudrama\u003c/font\u003e\u003c/p\u003e\u003c/div\u003e\u003cp dir\u003d\"ltr\"\u003e\u003cbr\u003e\u003c/p\u003e\u003cdiv align\u003d\"left\" \u003e\u003cp dir\u003d\"ltr\"\u003e\u003cfont color \u003d\"-16777216\"\u003eUntil the lights take us\u003c/font\u003e\u003c/p\u003e\u003c/div\u003e\u003cp dir\u003d\"ltr\"\u003e\u003cbr\u003e\u003cbr\u003e\u003c/p\u003e\u003cdiv align\u003d\"left\" \u003e\u003cp dir\u003d\"ltr\"\u003e\u003cfont color \u003d\"-16777216\"\u003eBeatles across \u003c/font\u003e\u003cfont color \u003d\"-16777216\"\u003ethe universe\u003c/font\u003e\u003c/p\u003e\u003c/div\u003e\u003cp dir\u003d\"ltr\"\u003e\u003cbr\u003e\u003c/p\u003e",
    "DrawImage": "",
    "ObjectOrder": "T",
    "PreviewImage": "",
    "ReminderText": "",
    "Color": -462108,
    "CategoryId": 1,
    "CreatedTime": 1474560571808,
    "Id": 48,
    "ModifiedTime": 1474560881071,
    "DeviceWidth": 540,
    "DrawLayoutHeight": 960,
    "FontSizePx": 27,
    "IsLocked": 0,
    "IsSynced": 2,
    "Order": 0,
    "Style": 0
  },
  "MemoObjectList": [
    {
      "Desc": "\u003cdiv align\u003d\"left\" \u003e\u003cp dir\u003d\"ltr\"\u003e\u003cfont color \u003d\"-16777216\"\u003eGroup mayhem\u003c/font\u003e\u003c/p\u003e\u003c/div\u003e\u003cp dir\u003d\"ltr\"\u003e\u003cbr\u003e\u003c/p\u003e\u003cdiv align\u003d\"left\" \u003e\u003cp dir\u003d\"ltr\"\u003e\u003cfont color \u003d\"-16777216\"\u003eInteresting docudrama\u003c/font\u003e\u003c/p\u003e\u003c/div\u003e\u003cp dir\u003d\"ltr\"\u003e\u003cbr\u003e\u003c/p\u003e\u003cdiv align\u003d\"left\" \u003e\u003cp dir\u003d\"ltr\"\u003e\u003cfont color \u003d\"-16777216\"\u003eUntil the lights take us\u003c/font\u003e\u003c/p\u003e\u003c/div\u003e\u003cp dir\u003d\"ltr\"\u003e\u003cbr\u003e\u003cbr\u003e\u003c/p\u003e\u003cdiv align\u003d\"left\" \u003e\u003cp dir\u003d\"ltr\"\u003e\u003cfont color \u003d\"-16777216\"\u003eBeatles across \u003c/font\u003e\u003cfont color \u003d\"-16777216\"\u003ethe universe\u003c/font\u003e\u003c/p\u003e\u003c/div\u003e\u003cp dir\u003d\"ltr\"\u003e\u003cbr\u003e\u003c/p\u003e",
      "DescRaw": "Group mayhem\nInteresting docudrama\nUntil the lights take us\n\nBeatles across the universe\n",
      "Id": 51,
      "MemoId": 48,
      "Alignment": 3,
      "Angle": 0,
      "Height": 197,
      "IsChecked": 0,
      "OrderNum": 0,
      "Type": 0,
      "Width": 504,
      "X": 18,
      "Y": 18
    }
  ]
}
END_JLQM
    $jlqm_contents;
}
1;


lib/LG/QuickMemo_Plus/Extract/Memo.pm
package LG::QuickMemo_Plus::Extract::Memo;
use 5.008001;
use strict;
use warnings;
use JSON;
use Archive::Zip qw( :ERROR_CODES :CONSTANTS );

our $VERSION = "0.01";

use Exporter qw(import);
 
our @EXPORT_OK = qw( lqm_to_str );
our $suppress_header = 0;

sub lqm_to_str {
    ## pass an lqm file exported from QuickMemo+
    my ( $lqm_file ) = @_;
    if (not -f $lqm_file){
        warn "$lqm_file is not a file";
        return '';
    }
    my $note_created_time = "";
    if ( $lqm_file =~ /(QuickMemo\+_(\d{6}_\d{6})(\(\d+\))?)/i) {
        $note_created_time = $2;
    }
    my $ref_json_str = extract_json_from_lqm( $lqm_file );
    return '' if not $ref_json_str;
    my ($extracted_text, $note_category) = extract_text_from_json($ref_json_str);
    my $header = "Created date: $note_created_time\n";
    $header .= "Category:   $note_category\n";
    $header .= "-"x79 . "\n";
    $header = '' if $suppress_header;
    return $header . $extracted_text;
}

#####################################
#         Unzip jlqm file and 
#           return json file contents.
#
sub extract_json_from_lqm {
    #   unzip
    #   extract the memoinfo.jlqm file.
    my $lqm_file = shift;
    # Read a Zip file
    my $lqm_zip = Archive::Zip->new();
    unless ( $lqm_zip->read( $lqm_file ) == AZ_OK ) {
        warn "Error reading $lqm_file";
        ####### to do: add the zip error to the warning?
        return "";
    }
    my $jlqm_filename = "memoinfo.jlqm";
    my $member = $lqm_zip->memberNamed( $jlqm_filename );
    ############### to do: add warning here if memoinfo.jlqm is missing.
    if( not $member ){
        warn "File not found: $jlqm_filename in archive $lqm_file";
        return "";
    }
    my ( $string, $status ) = $member->contents();
    if(not $status == AZ_OK){
        warn "Error extracting $jlqm_filename from $lqm_file : Status = $status";
        return "";
    }
    return \$string;
}

###############################################
#       Decode json file contents and
#         return the text in 'DescRaw'
sub extract_text_from_json {
    my $ref_json_string = shift;
    
    ############# To do: eval this and trap errors.
    my $href_memo  = decode_json $$ref_json_string;
    if (not $href_memo){
        warn "Error decoding text";
        return '','';
    }
    my $text = "";
    foreach( @{$href_memo->{MemoObjectList}} ) {
        $text .= $_->{DescRaw};
        $text .= "\n";
    }
    my $category = $href_memo->{Category}->{CategoryName};
    $category //= '';
    $category =~ s/[^\w-]/_/;
    return $text, $category;
}
1;
__END__

=encoding utf-8

=head1 NAME

LG::QuickMemo_Plus::Extract::Memo

=head1 SYNOPSIS

    use LG::QuickMemo_Plus::Extract::Memo qw(lqm_to_str);
    my $memo_text = lqm_to_str('QuickMemo+_191208_220400.lqm');

=head1 DESCRIPTION

LG::QuickMemo_Plus::Extract::Memo is a module that will extract the 
text contents from archived QuickMemo+ memos.

QuickMemo+ lqm files are in Zip format. This program unzips them, 
parses the json file inside, then extracts the category and memo text 
from the Json file.

=head1 LICENSE

Copyright (C) Brent Shields.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Brent Shields E<lt>bshields@cpan.orgE<gt>

=cut



lib/Data/QuickMemoPlus/Reader.pm
package Data::QuickMemoPlus::Reader;
use 5.010;
use strict;
use warnings;
use Carp;
use JSON;
use Archive::Zip qw( :ERROR_CODES :CONSTANTS );

our $VERSION = "0.11";

use Exporter qw(import);
 
our @EXPORT_OK = qw( lqm_to_str lqm_to_txt );
our $IncludeHeader = 1;

sub lqm_to_txt {
    my ( $lqm ) = @_;
    my $files_converted = 0;
    if (-d $lqm){
        $lqm =~ s![/\\]+$!!;
        foreach(glob qq("$lqm/*.lqm")){
            $files_converted += _lqm_file_to_txt($_);
        }
    } else {
        $files_converted = _lqm_file_to_txt($lqm);
    }
    return $files_converted;
}
sub _lqm_file_to_txt {
    my ( $lqm_file ) = @_;
    my $text = lqm_to_str($lqm_file);
    
    return 0 if not length($text);
    
    my $outfile = $lqm_file;
    $outfile =~ s/lqm$/txt/i;
    open my $fh, ">", $outfile or do {
        carp "couldn't open $outfile\n";
        
        return 0;
    };
    print $fh $text;
    
    return 1;
}

sub lqm_to_str {
    my ( $lqm_file ) = @_;
    if (not -f $lqm_file){
        carp "$lqm_file is not a file";
        
        return '';
    }
    my $note_created_time = "";
    if ( $lqm_file =~ /(QuickMemo\+_(\d{6}_\d{6})(\(\d+\))?)/i) {
        $note_created_time = $2;
    }
    my $json_str = extract_json_from_lqm( $lqm_file );
    
    return '' if not $json_str;
    
    my ($extracted_text, $note_category) = extract_text_from_json($json_str);
    my $header = "Created date: $note_created_time\n";
    $header .= "Category:   $note_category\n";
    $header .= "-"x79 . "\n";
    $header = '' if not $IncludeHeader;
    
    return $header . $extracted_text;
}

#####################################
#
sub extract_json_from_lqm {
    my $lqm_file = shift;
    my $lqm_zip = Archive::Zip->new();
    unless ( $lqm_zip->read( $lqm_file ) == AZ_OK ) {
        carp "Error reading $lqm_file";
        ####### to do: add the zip error to the warning?
        
        return "";
    }
    my $jlqm_filename = "memoinfo.jlqm";
    my $member = $lqm_zip->memberNamed( $jlqm_filename );
    if( not $member ){
        carp "File not found: $jlqm_filename in archive $lqm_file";
        
        return "";
    }
    my ( $string, $status ) = $member->contents();
    if(not $status == AZ_OK){
        carp "Error extracting $jlqm_filename from $lqm_file : Status = $status";
        
        return "";
    }
    
    return $string;
}

###############################################
#
sub extract_text_from_json {
    my $json_string = shift;
    
    ############# To do: eval this and trap errors.
    my $href_memo  = decode_json $json_string;
    if (not $href_memo){
        carp "Error decoding JSON file in lqm archive.";
        return '','';
    }
    my $text = "";
    foreach( @{$href_memo->{MemoObjectList}} ) {
        $text .= $_->{DescRaw};
        $text .= "\n";
    }
    my $category = $href_memo->{Category}->{CategoryName};
    $category //= '';
    $category =~ s/[^\w-]/_/g;
    return $text, $category;
}
1;
__END__

=encoding utf-8

=head1 NAME

Data::QuickMemoPlus::Reader - Extract text from QuickMemo+ LQM export files.

=head1 SYNOPSIS

    use Data::QuickMemoPlus::Reader qw(lqm_to_str);
    my $memo_text = lqm_to_str('QuickMemo+_191208_220400.lqm');

    use Data::QuickMemoPlus::Reader qw(lqm_to_txt);
    my $files_converted1 = lqm_to_txt('QuickMemo+_191208_220400.lqm');
    my $files_converted2 = lqm_to_txt('path/to/lqm_files');
    
    ## Omit the header text by setting setting this package variable to false:
    local $Data::QuickMemoPlus::Reader::IncludeHeader;

=head1 DESCRIPTION

C<Data::QuickMemoPlus::Reader> is a module that will extract the 
text contents from archived QuickMemo+ memos. QuickMemo+ is a memo 
application that comes with LG smartphones.

QuickMemo+ F<lqm> files are in Zip format. This module unzips them, 
parses the json file inside, then extracts the category and memo text 
from the Json file.

If the filename of the lqm file contains the original timestamp then that
is placed in a text header in the text along with the category name. The header
can be disabled by setting the package variable C<$IncludeHeader> to false.

The following functions are available:

=head2 lqm_to_txt('directory or filename')

Creates a text file with the same name as each original lqm file but with a txt extension.
Return value is the number of files successfully converted.

=head2 lqm_to_str('filename')

Returns the text extracted from the lqm file.

=head1 LICENSE

Copyright (C) Brent Shields.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Brent Shields E<lt>bshields@cpan.orgE<gt>

=cut

